### PR TITLE
fs: add uv_fs_mkstemp()

### PR DESCRIPTION
**Pull request by bshm**
Credit goes to Andreas Hauptmann (@fdopen) for most of the Windows part and to Florian Margaine (@ralt) for the rest. This is a follow-up of #729.
The Windows implementation extends existing code for mkdtemp.
For Unix OSs mkostemp64, mkostemp and mkstemp are tried in order to use O_CLOEXEC.

### CODE REVIEW

**Comment by bshm** [comment]
Follow-up of #729

File: src/unix/core.c [hunk 1]
```diff
@@ -79,6 +79,10 @@
 # endif
 #endif
 
+#ifdef _AIX
+#include <sys/ioctl.h>
```
Comment (richardlau) [inline src/unix/core.c]: Is this necessary? It appears to be unconditionally included in L34 above.

Comment (bshm) [reply src/unix/core.c]: removed unnecessary include statement

**Comment by bshm** [comment]
@richardlau could you please continue the review

**Comment by musm** [comment]
Any chance we could push this through soon?

File: test/test-fs.c [hunk 2]
```diff
@@ -71,6 +71,7 @@ static int write_cb_count;
 static int unlink_cb_count;
```
Comment (cjihrig) [inline test/test-fs.c]: In this same file, can you update the `fs_null_req` test to also call `uv_fs_mkstemp()` with `NULL` arguments.

File: test/test-fs.c [hunk 3]
```diff
@@ -104,6 +105,9 @@ static uv_fs_t close_req;
 static uv_fs_t mkdir_req;
 static uv_fs_t mkdtemp_req1;
 static uv_fs_t mkdtemp_req2;
+static uv_fs_t mkstemp_req1;
+static uv_fs_t mkstemp_req2;
+static uv_fs_t mkstemp_req3;
```
Comment (cjihrig) [inline test/test-fs.c]: Small nit, but I think you could reorganize the test code a bit to only use one or two `mkstemp_req` variables.

File: src/unix/fs.c [hunk 4]
```diff
@@ -1371,6 +1459,21 @@ int uv_fs_mkdtemp(uv_loop_t* loop,
 }
 
 
+int uv_fs_mkstemp(uv_loop_t* loop,
+                  uv_fs_t* req,
+                  const char* tpl,
+                  uv_fs_cb cb) {
+  INIT(MKSTEMP);
+  req->path = uv__strdup(tpl);
```
Comment (saghul) [inline src/unix/fs.c]: don't we now have PATH (the macro) doing all this?

**Comment by musm** [comment]
bump

File: src/unix/fs.c [hunk 5]
```diff
@@ -220,6 +220,93 @@ static ssize_t uv__fs_mkdtemp(uv_fs_t* req) {
 }
 
 
+static int uv__fs_mkstemp(uv_fs_t* req) {
+  int r;
+#ifdef UV__O_CLOEXEC
```
Comment (saghul) [inline src/unix/fs.c]: This needs a rebase, the logic to detect O_CLOEXEC got simplified in https://github.com/libuv/libuv/pull/2454

File: src/unix/fs.c [hunk 6]
```diff
@@ -220,6 +220,93 @@ static ssize_t uv__fs_mkdtemp(uv_fs_t* req) {
...
+  path = (char*) req->path;
+  path_length = strlen(path);
+
+  /* EINVAL can be returned for 2 reasons:
+       1. The template's last 6 characters were not XXXXXX
+       2. open() didn't support O_CLOEXEC
+     We want to avoid going to the fallback path in case
+     of 1., so it's manually checked before. */
+  if (path_length < strlen(pattern) ||
+      strcmp(path + (path_length - (sizeof(pattern) - 1)), pattern)) {
```
Comment (saghul) [inline src/unix/fs.c]: here we are computing the length of `pattern` twice, once with strlen (unnecessary) and another one with sizeof - 1. Please either make it a constant or use sizeof on both cases for consistency.

File: src/unix/fs.c [hunk 7]
```diff
@@ -1371,6 +1459,21 @@ int uv_fs_mkdtemp(uv_loop_t* loop,
...
 
+int uv_fs_mkstemp(uv_loop_t* loop,
+                  uv_fs_t* req,
+                  const char* tpl,
+                  uv_fs_cb cb) {
+  INIT(MKSTEMP);
+  req->path = uv__strdup(tpl);
+  if (req->path == NULL) {
+    if (cb != NULL)
+      uv__req_unregister(loop, req);
```
Comment (saghul) [inline src/unix/fs.c]: no need to do this, cleanup is expected from the library user

File: src/unix/internal.h [hunk 8]
```diff
@@ -48,6 +48,22 @@
...
 #endif /* __sun */
 
+#if defined(__FreeBSD__) || defined(__DragonFly__)
+# include <sys/sysctl.h>
+# include <sys/filio.h>
+# include <sys/ioctl.h>
+# include <sys/wait.h>
+# define UV__O_CLOEXEC O_CLOEXEC
+# if defined(__FreeBSD__) && __FreeBSD__ >= 10
+#  define uv__accept4 accept4
```
Comment (saghul) [inline src/unix/internal.h]: why is all of this needed? seems unrelated

**Comment by saghul** [comment]
@bshm Sorry this fell through the cracks. Any chance you can pick it up? If not (that's perfectly fine), would you mind if someone else continued where you left off?

**Comment by musm** [comment]
Looking forward to this functionality

**Comment by musm** [comment]
Note this PR closes https://github.com/libuv/libuv/issues/2555 , any chance we could see it updated

**Comment by saghul** [comment]
Closing in favor of https://github.com/libuv/libuv/pull/2557 @musm Feel free to give it a try and provide feedback.
